File: redhawk-codegen/redhawk/codegen/jinja/python/component/persona/templates/persona_base.py
#{#
# This file is protected by Copyright. Please refer to the COPYRIGHT file
# distributed with this source distribution.
#
# This file is part of REDHAWK core.
#
# REDHAWK core is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# REDHAWK core is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.
#}
#% set className = component.personaclass.name
#% set baseClass = component.baseclass.name
#% set artifactType = component.artifacttype

import os
import sys
from ossie.cf import CF
#{% if component is device %}
from ossie.device import start_device
#{% else %}
from ossie.resource import Resource, start_component
#{% endif %}
import logging

from ${baseClass} import *


class ${className}(${baseClass}):

    def initialize(self):
        ${baseClass}.initialize(self)
        self._parentDevice = None
        self._parentAllocated = False
        self._previousRequestProps = None
#{% if isExecutable == True %}
        self._resourceMap = {}
        self._processMap = {}
        self._processIdIncrement = 0
#{% endif %}

    def releaseObject(self):
#{% if isExecutable == True %}
        for pid in self._processMap:
            self.terminate(pid)

#{% endif %}
        self._baseLog.debug("releaseObject()")
        if self._adminState == CF.Device.UNLOCKED:
            self._adminState = CF.Device.SHUTTING_DOWN

        try:
            self._unregister()
        except Exception, e:
            raise CF.LifeCycle.ReleaseError(str(e))

        self._adminState = CF.Device.LOCKED
        try:
            objid = self._default_POA().servant_to_id(self)
            self._default_POA().deactivate_object(objid)
        except:
            self._baseLog.error("failed releaseObject()")

    def attemptToProgramParent(self):
        if (not self._parentDevice):
            self._baseLog.error("Unable to Program parent: No reference to parent device exists!")
            return False

        if (not self._parentAllocated):
            self._baseLog.debug("About to allocate parent device!")

            self.beforeHardwareProgrammed()
            requestProps   = self.hwLoadRequest()
            formattedProps = self._formatRequestProps(requestProps)

            if not formattedProps:
                self._baseLog.error("Failed to format hw_load_request props for parent device")
                return False

            self._parentAllocated = self._parentDevice.allocateCapacity(formattedProps)

            if (self._parentAllocated):
                self._previousRequestProps = formattedProps
                self.afterHardwareProgramSuccess()
            else:
                self.afterHardwareProgramFailure()

        return self._parentAllocated

    def attemptToUnprogramParent(self):
        if (not self._parentDevice):
            self._baseLog.error("Unable to Program parent: No reference to parent device exists!")
            return False

        if self._parentAllocated:
            self._baseLog.debug("About to deallocate parent device!")

            if (not self._previousRequestProps):
                self._baseLog.error("Previously requested hw_load props empty!")
                return False

            self._parentDevice.deallocateCapacity(self._previousRequestProps)
            self._parentAllocated = False
            self.afterHardwareUnprogrammed();

        return (not self._parentAllocated)

#{% if isExecutable == True %}
    def execute(self, name, options, parameters):
        
        for param in parameters:
            propId = param.id
            propVal = param.value.value()
            if type(propVal) == str:
                self._baseLog.debug("InstantiateResourceProp: ID['" + 
                    str(propId) + "'] = " + str(propVal))

        resource = instantiateResource(name, options, parameters)
        
        if not resource:
            msg = "Unable to dynamically instantiate resource!"
            self._baseLog.error(msg)
            raise CF.ExecutableDevice.ExecuteFail(CF.CF_NOTSET, msg)

        resourceId = resource._get_identifier()
        fakePid = self._processIdIncrement + 1
        self._resourceMap[resourceId] = resource
        self._processMap[fakePid] = resourceId
        self._processIdIncrement += 1

        return fakePid

    def terminate(self, processId):
        if not self._processMap.has_key(processId):
            raise CF.ExecutableDevice.InvalidProcess(CF.CF_ENOENT,
                "Cannot terminate.  Process %s does not reference a resource!." % str(processId))

        resourceId = self._processMap.pop(processId)

        if not self._resourceMap.has_key(resourceId):
            raise CF.ExecutableDevice.InvalidState(
                "Cannot terminate.  Unable to locate resource '%s'!" % str(resourceId))

        resource = self._resourceMap.pop(resourceId)
        del resource

    
    def instantiateResource(self, libraryName, options, parameters):
        resource = None
        # TODO fill this in
        return resource
#{% endif %}

    def allocateCapacity(self, capacities):
        allocationSuccess = ${baseClass}.allocateCapacity(self, capacities)
        if allocationSuccess: 
            allocationSuccess = self.attemptToProgramParent()
            # Undo current allocations if we can't program parent
            if not allocationSuccess:
                ${baseClass}.deallocateCapacity(self, capacities)
        return allocationSuccess

    def deallocateCapacity(self, capacities):
        ${baseClass}.deallocateCapacity(self, capacities)
#{% if isExecutable == True %}
        if not self.hasRunningResource():
            self.attemptToUnprogramParent()
#{% else %}
        self.attemptToUnprogramParent()
#{% endif %}

    def hwLoadRequest(self):
        return []

    def beforeHardwareProgrammed(self):
        # User may override this method
        return

    def beforeHardwareUnprogrammed(self): 
        # User may override this method
        return

    def afterHardwareProgramSuccess(self): 
        # User may override this method
        return

    def afterHardwareProgramFailure(self): 
        # User may override this method
        return

    def afterHardwareUnprogrammed(self): 
        # User may override this method
        return

    def _formatRequestProps(self, requestProps):
        # Sanity check... Can't format nothing!
        if not requestProps:
            self._baseLog.error("Unable to format hw_load_request_properties.  Properties are empty!")
            return None

        # Sanity check... Make sure the type has an id field!
        if not hasattr(requestProps[0], "id"):
            self._baseLog.error("Unable to format hw_load_request_properties.  Properties must be of list of 'CF.Datatype'")
            return None

        # Case 1 - Properties are already formatted properly
        if len(requestProps) == 1:
            if requestProps[0].id == "hw_load_requests":
                self._baseLog.debug("No formatting occurred - Assumed formatting is proper")
                return requestProps

        # Further inspection of properties
        allPropsAreHwLoadRequest = True
        foundRequestId = False
        for prop in requestProps:
            allPropsAreHwLoadRequest &= (prop.id == "hw_load_request")
            foundRequestId |= (prop.id == "request_id") or \
                              (prop.id == "hw_load_request::request_id")

        # Case 2 - Properties are list of hw_load_request structs
        if allPropsAreHwLoadRequest:
            self._baseLog.debug("Found hw_load_request list - Formatting to structseq")
            return [CF.DataType(id="hw_load_requests", value=requestProps)]

        # Case 3 - Properties represent the contents of a single hw_load_request
        if foundRequestId:
            self._baseLog.debug("Found hw_load_request contents - Formatting to structseq")
            structProp = CF.DataType(id="hw_load_request", value=requestProps)
            return [CF.DataType(id="hw_load_requests", value=structProp)]

        self._baseLog.error("Unable to format hw_load_request_properties.  Properties are empty!")
        return None

File: redhawk-codegen/redhawk/codegen/jinja/cpp/component/persona/templates/persona_base.cpp
/*#
 * This file is protected by Copyright. Please refer to the COPYRIGHT file 
 * distributed with this source distribution.
 * 
 * This file is part of REDHAWK core.
 * 
 * REDHAWK core is free software: you can redistribute it and/or modify it 
 * under the terms of the GNU Lesser General Public License as published by the 
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * REDHAWK core is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License 
 * for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License 
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 #*/
/*{% block license %}*/
/*# Allow child templates to include license #*/
/*{% endblock %}*/
//% set className = component.reprogclass.name
//% set baseClass = component.baseclass.name
//% set artifactType = component.artifacttype
/**************************************************************************

    This is the ${artifactType} code. This file contains the child class where
    custom functionality can be added to the ${artifactType}. Custom
    functionality to the base class can be extended here. Access to
    the ports can also be done from this class

**************************************************************************/

#include "${component.reprogclass.header}"

PREPARE_LOGGING(${className})

/*{% if component is device %}*/
${className}::${className} (
                        char            *devMgr_ior, 
                        char            *id, 
                        char            *lbl, 
                        char            *sftwrPrfl ) 
:
    ${baseClass}(devMgr_ior, id, lbl, sftwrPrfl),
    _previousRequestProps()
{
    construct();
}

${className}::${className} (
                        char            *devMgr_ior, 
                        char            *id, 
                        char            *lbl, 
                        char            *sftwrPrfl, 
                        char            *compDev ) 
:
    ${baseClass}(devMgr_ior, id, lbl, sftwrPrfl, compDev),
    _previousRequestProps()
{
    construct();
}

${className}::${className} (
                        char            *devMgr_ior, 
                        char            *id, 
                        char            *lbl, 
                        char            *sftwrPrfl, 
                        CF::Properties  capacities ) 
:
    ${baseClass}(devMgr_ior, id, lbl, sftwrPrfl, capacities),
    _previousRequestProps()
{
    construct();
}

${className}::${className} (
                        char            *devMgr_ior, 
                        char            *id, 
                        char            *lbl, 
                        char            *sftwrPrfl, 
                        CF::Properties  capacities, 
                        char            *compDev ) 
:
    ${baseClass}(devMgr_ior, id, lbl, sftwrPrfl, capacities, compDev),
    _previousRequestProps()
{
    construct();
}
/*{% else %}*/
${className}::${className} (
                        const char      *uuid, 
                        const char      *label ) 
:
    ${baseClass}(uuid, label),
    _previousRequestProps()
{
    construct();
}
/*{% endif %}*/

void ${className}::construct() 
{
    // Initialize state to safe defaults
    _parentDevice = NULL;
    _parentAllocated = false;
//% if component is executabledevice
    _processIdIncrement = 0;
//% endif 
}

// TODO: This was overriden since setting admin state is not accessible via the current IDL
void ${className}::adminState(CF::Device::AdminType adminState) 
    throw (CORBA::SystemException)
{
    // Force admin state to change usage state since usage state is currently protected
    switch (adminState) {
        case CF::Device::LOCKED:
            setUsageState(CF::Device::BUSY);
            break;
        case CF::Device::UNLOCKED:
            setUsageState(CF::Device::IDLE);
            break;
        case CF::Device::SHUTTING_DOWN:
            // Do nothing
            break;
    }

    ${baseClass}::adminState(adminState);
}

void ${className}::releaseObject() 
    throw (
        CF::LifeCycle::ReleaseError, 
        CORBA::SystemException ) 
{
/*{% if component is executabledevice %}*/
    // Terminate all children that were executed
    ProcessMapIter iter;
    for (iter = _processMap.begin(); iter != _processMap.end(); iter++) {
        this->terminate(iter->first);
    }
/*{% endif %}*/
    // deactivate ports
    releaseInPorts();
    releaseOutPorts();

    // SR:419
    RH_DEBUG(this->_deviceLog, __FUNCTION__ << ": Receive releaseObject call");
    if (_adminState == CF::Device::UNLOCKED) {
        RH_DEBUG(this->_deviceLog, __FUNCTION__ << ": Releasing Device")
        setAdminState(CF::Device::SHUTTING_DOWN);

        // SR:418
        // TODO Release aggregate devices if more than one exists
        if (!CORBA::is_nil(_aggregateDevice)) {
            try {
                _aggregateDevice->removeDevice(this->_this());
            } catch (...) {
            }
        }

        RH_DEBUG(this->_deviceLog, __FUNCTION__ << ": Done Releasing Device")
    }
}


CORBA::Boolean ${className}::attemptToProgramParent() 
{
    // Return false if there is no reference to the parent
    if (_parentDevice == NULL) {
        RH_ERROR(this->_deviceLog, __FUNCTION__ << 
            ": No reference to parent exists!");
        return false;
    }

    if (_parentAllocated == false) {

        RH_DEBUG(this->_deviceLog, __FUNCTION__ << 
            ": About to allocate parent device");
        
        beforeHardwareProgrammed();

        // Grab user-defined allocation request and format them properly
        CF::Properties requestProps;
        CF::Properties formattedRequestProps;

        hwLoadRequest(requestProps);
        formatRequestProps(requestProps, formattedRequestProps);

        _parentAllocated = _parentDevice->allocateCapacity(formattedRequestProps);

        if (_parentAllocated) {
            _previousRequestProps = formattedRequestProps;
            afterHardwareProgramSuccess();
        } else {
            afterHardwareProgramFailure();
        }
    }
    return _parentAllocated;
}

CORBA::Boolean ${className}::attemptToUnprogramParent() 
{
    // Return false if there is no reference to the parent
    if (_parentDevice == NULL) {
        RH_ERROR(this->_deviceLog, __FUNCTION__ << ": No reference to parent exists!");
        return false;
    }
    
    if (_parentAllocated) {
        beforeHardwareUnprogrammed();
        
        // Grab previous user-defined allocation request
        if (_previousRequestProps.length() == 0) {
            RH_ERROR(this->_deviceLog, __FUNCTION__ << ": Previously requested hw_load Props empty!");
            return false;
        }

        _parentDevice->deallocateCapacity(_previousRequestProps);
        _parentAllocated = false;
        afterHardwareUnprogrammed();
    }
    return !_parentAllocated;
}

/*{% if component is executabledevice %}*/
CF::ExecutableDevice::ProcessID_Type ${className}::execute (
                        const char*                 name, 
                        const CF::Properties&       options, 
                        const CF::Properties&       parameters )
    throw ( 
        CF::ExecutableDevice::ExecuteFail, 
        CF::InvalidFileName, 
        CF::ExecutableDevice::InvalidOptions, 
        CF::ExecutableDevice::InvalidParameters,
        CF::ExecutableDevice::InvalidFunction, 
        CF::Device::InvalidState, 
        CORBA::SystemException ) 
{
    // Initialize local variables
    std::string propId;
    std::string propValue;
    std::string resourceId;
    Resource_impl* resourcePtr = NULL;
    
    // Iterate through all parameters for debugging purposes
    for (unsigned int ii = 0; ii < parameters.length(); ii++) {
        propId = parameters[ii].id;
        propValue = ossie::any_to_string(parameters[ii].value);
        RH_DEBUG(this->_deviceLog, __FUNCTION__ << 
            ": InstantiateResourceProp: ID['" << propId << "'] = " << propValue);
    }

    // Attempt to create and verify the resource
    resourcePtr = instantiateResource(name, options, parameters);
    if (resourcePtr == NULL) {
        RH_FATAL(this->_deviceLog, __FUNCTION__ << 
            ": Unable to instantiate '" << name << "'");
        throw (CF::ExecutableDevice::ExecuteFail());
    }

    resourceId = ossie::corba::returnString(resourcePtr->identifier());
    _resourceMap[resourceId] = resourcePtr;                 // Store the resourcePtr
    _processMap[++_processIdIncrement] = resourceId;        // Store the resourcePtr Process for termination
    //_deviceManager->registerDevice(resourcePtr->_this());

    return (CORBA::Long) _processIdIncrement;
}

void ${className}::terminate(CF::ExecutableDevice::ProcessID_Type processId)
    throw (
        CF::Device::InvalidState, 
        CF::ExecutableDevice::InvalidProcess, 
        CORBA::SystemException ) 
{
    // Initialize local variables
    ProcessMapIter processIter;
    ResourceMapIter resourceIter;
    ResourceId resourceId;

    // Search for the resourceId that's related to the incoming terminate request
    processIter = _processMap.find(processId);
    if (processIter != _processMap.end()) {

        /// Search for the persona that related to the found resourceId
        resourceIter = _resourceMap.find(processIter->second);
        if (resourceIter != _resourceMap.end()) {
            _processMap.erase(processIter);
            _resourceMap.erase(resourceIter);

            // We don't need to call releaseObject here since HW Components will
            // be released by the application factory
            delete resourceIter->second;

            return;
        }
    }
}

bool ${className}::hasRunningResources() 
{
    return (!_resourceMap.empty());
}

Resource_impl* ${className}::instantiateResource(
                        const char*                 libraryName, 
                        const CF::Properties&       options, 
                        const CF::Properties&       parameters) 
{
    // Initialize local variables
    std::string absPath = get_current_dir_name();
    void* pHandle = NULL;
    char* errorMsg = NULL;
    CF::Properties combinedProps;
    unsigned int skipRunInd = 0;
   
    std::string propId;
    std::string propValue;

    const char* symbol = "construct";
    void* fnPtr = NULL; 
    unsigned long argc = 0;
    unsigned int argCounter = 0;
    
    ConstructorPtr constructorPtr = NULL;
    Resource_impl* resourcePtr = NULL;


    // Open up the cached .so file
    absPath.append(libraryName);
    pHandle = dlopen(absPath.c_str(), RTLD_NOW);
    if (!pHandle) {
        errorMsg = dlerror();
        RH_FATAL(this->_deviceLog, __FUNCTION__ <<  
                ": Unable to open library '" << absPath.c_str() << "': " << errorMsg);
        return NULL;
    }

    // Add SKIP_FLAG to properties
    combinedProps = parameters;
    skipRunInd = combinedProps.length();
    combinedProps.length(skipRunInd + 1);
    combinedProps[skipRunInd].id = CORBA::string_dup("SKIP_RUN");
    combinedProps[skipRunInd].value <<= true;

    // Convert combined properties into ARGV/ARGC format
    argc = combinedProps.length() * 2;
    char* argv[argc];
    for (unsigned int i = 0; i < combinedProps.length(); i++) {
        propId = combinedProps[i].id;
        propValue = ossie::any_to_string(combinedProps[i].value);

        argv[argCounter] = (char*) malloc(propId.size() + 1);
        strcpy(argv[argCounter++], propId.c_str());

        argv[argCounter] = (char*) malloc(propValue.size() + 1);
        strcpy(argv[argCounter++], propValue.c_str());
    }

    // Look for the 'construct' C-method
    fnPtr = dlsym(pHandle, symbol);
    if (!fnPtr) {
        errorMsg = dlerror();
        RH_FATAL(this->_deviceLog, __FUNCTION__ << 
            ": Unable to find symbol '" << symbol << "': " << errorMsg);
        return NULL;
    }

    // Cast the symbol as a ConstructorPtr
    constructorPtr = reinterpret_cast<ConstructorPtr>(fnPtr);
    
    // Attempt to instantiate the resource via the constructor pointer
    try {
        resourcePtr = generateResource(argc, argv, constructorPtr, libraryName);
    } catch (...) {
        RH_FATAL(this->_deviceLog, __FUNCTION__ << 
            ": Unable to construct persona device: '" << argv[0] << "'");
    }

    // Free the memory used to create argv
    for (unsigned int i = 0; i < argCounter; i++) {
        free(argv[i]);
    }

    return resourcePtr;
}
/*{% endif %}*/

// Transforms user-supplied properties into safe-usable CF::Properties
void ${className}::formatRequestProps(
                    const CF::Properties&       requestProps, 
                    CF::Properties&             formattedProps) 
{
    // Initialize local variables
    std::string propId;
    bool allPropsAreHwLoadRequest = false;
    bool foundRequestId = false;
    CF::Properties hwLoadRequest;

    // Sanity check - Kick out if properties are empty
    if (requestProps.length() == 0) {
        RH_ERROR(this->_deviceLog, __FUNCTION__ << 
            ": Unable to format hw_load_request properties.  Properties are empty!");
        return;
    }

    // Case 1 - Properties are already properly formatted
    if (requestProps.length() == 1) {
        propId = requestProps[0].id;
        if (propId == "hw_load_requests") {
            RH_DEBUG(this->_deviceLog, __FUNCTION__ <<
                ": No formatting occurred - Request properties are properly formatted!");
            formattedProps = requestProps;
            return;
        }
    }

    // Inspect the properties further
    allPropsAreHwLoadRequest = true;
    for (size_t ii = 0; ii < requestProps.length(); ii++) {
        propId = requestProps[ii].id;

        allPropsAreHwLoadRequest &= (propId == "hw_load_request");
        foundRequestId    |= (propId == "request_id") ||
                             (propId == "hw_load_request::request_id");
    }
  
    // Case 2 - Properties are multiple hw_load_request structs
    if (allPropsAreHwLoadRequest) {
        RH_DEBUG(this->_deviceLog, __FUNCTION__ << 
            ": Found hw_load_request array - Formatting to structseq");
        formattedProps.length(1);
        formattedProps[0].id = "hw_load_requests";
        formattedProps[0].value <<= requestProps;
        return;
    }

    // Case 3 - Properties reprensent the contents of a single hw_load_request
    if (foundRequestId) {
        RH_DEBUG(this->_deviceLog, __FUNCTION__ <<
            ": Found hw_load_request contents - Formatting to struct and structseq");
        
        hwLoadRequest.length(1);
        hwLoadRequest[0].id = "hw_load_request";
        hwLoadRequest[0].value <<= requestProps;

        formattedProps.length(1);
        formattedProps[0].id = "hw_load_requests";
        formattedProps[0].value <<= hwLoadRequest;
        return;
    }
    
    RH_ERROR(this->_deviceLog, __FUNCTION__ <<
        ": Unable to format hw_load_request properties - Format unknown!");
}
